### DISCUSSION TITLE
Node.js for Android - All in One or Separated?

### DISCUSSION THREAD

**Discussion by caoccao** [caoccao/Javet#304] [Polls]
Node.js for Android will be supported in v3.1.0. There are 2 options in terms of how to organize the AAR files.

## 1. All in One
 
Both Node.js and V8 will be packaged in the same AAR file.

### PROS

- There is only one dependency in the gradle file.

### CONS

- The AAR file size is doubled.
- Explicit `exclude` needs to be put in the gradle file.

## 2. Separated

Node.js and V8 will be packaged in separate AAR files.

### PROS

- The AAR file size is as usual.
- There is no need to write explicit `exclude` in the gradle file.

### CONS

- There are two dependencies in the gradle file.

## Note

- Application can only load either Node.js or V8 mode, but not both. The dynamic mode switch is not implemented on Android.

**Comment by jcrabanal** [comment]
Any attempt at modularization is always welcome. Just like when google modularized their colossal play services library years ago, and everyone instantly benefited from that.

And I'm thinking if javet ever supports i18n, it could be done the same way, so the people who never needed it will not be bothered in an javet upgrade.

I'd even welcome a callback or an option to load javet native .so libraries, and have them loaded by myself. When apps are not distributed by play store (where you can use android bundles for smaller size), you usually need to create a huge universal apk containing code for every platform.

There are already ways to do this (remove .so files from .aar, download manually the corresponding .so for the architecture of the device, then do System.loadLibrary() before touching any of the javet classes), but official support could be nice.

**Reply by caoccao** [reply]
Thank you for the advice. It makes sense.

Actually, all the features you mentioned have been available for years in Javet on all platforms except for Android. On Android, as a simple gradle `exclude` can work, so far I'm not going to release a permutation of those option. That applies to i18n. I've been under heavy burden in terms of preparing a new release. I wouldn't want my workload doubled or tripled by that. Hope that makes sense.

**Comment by jcrabanal** [comment]
Here's a proposal:

implementation("com.caoccao.javet:javet-android-base:3.0.4")
implementation("com.caoccao.javet:javet-android-base-native-libs:3.0.4")
implementation("com.caoccao.javet:javet-android-nodejs:3.0.4")
implementation("com.caoccao.javet:javet-android-i18n:3.0.4")

javet-android-base would be mandatory for everyone, and base-native-libs for most people too, and you'd have nodejs and a hypotetical i18n optional. This way, if some other optional future feature that requires huge files becomes a possibility, you could add it this way painlessly.
